cpp/include/cudf/join.hpp

```diff
@@ -958,8 +955,7 @@ std::unique_ptr<rmm::device_uvector<size_type>> mixed_left_semi_join(
   table_view const& left_conditional,
   table_view const& right_conditional,
   ast::expression const& binary_predicate,
-  null_equality compare_nulls = null_equality::EQUAL,
-  std::optional<std::pair<std::size_t, device_span<size_type const>>> output_size_data = {},
```

Comment:
This makes the PR a breaking change. 

We should keep the optional size input as it is since users may know the exact join output size in advance. Forcing the implementation to always reserve the upper bound size for output will increase the peak memory usage in this case.

Reply:
The current implementation will need to reserve the max output size even if the user provides the exact output data size. I can revert this change to not make it breaking, but I am not sure if we should keep this param if it does not affect function execution

Reply:
I'm supportive of the breaking change here. I don't think we need to worry about the potential increase in peak memory usage - we usually only worry about that for "data" and not a column of indices/bools.

Reply:
Peak memory usage for a semi/anti join typically isn't a huge issue since the max size is bounded by N rather than N x M. Also since this is consistent with the corresponding non-mixed API ([`left_semi_anti_join`](https://github.com/rapidsai/cudf/blob/c2bb860e4b323d1f9efd593938fef3372f36bdef/cpp/src/join/semi_join.cu#L44)) I think it's fine to change here.

Reply:
> The current implementation will need to reserve the max output size even if the user provides the exact output data size.

Valid point. 


Thank you all for your input. Consider me convinced.